onnxruntime/core/providers/cpu/math/element_wise_ops.h

```diff
@@ -6,132 +6,246 @@
 #include "core/common/common.h"
 #include "core/framework/op_kernel.h"
 #include "core/util/math_cpuonly.h"
+#include "core/providers/cpu/element_wise_ranged_transform.h"
 
 namespace onnxruntime {
+namespace functors {
 
-template <typename T>
-class Add final : public OpKernel {
- public:
-  Add(const OpKernelInfo& info) : OpKernel(info) {
+template<typename T>
+struct Log : public ElementWiseRangedTransform<T> {
+  Status Init(const onnxruntime::NodeAttributes) {
+    return Status::OK();
   }
 
-  Status Compute(OpKernelContext* context) const override;
+  ElementWiseRangedTransform<T>* Copy() const {
+    using T1 = typename std::remove_pointer<decltype(this)>::type;
+    using T2 = typename std::remove_const<T1>::type;
+    return new T2(*this);
+  }
+  
+  float Cost() const final { return 15.0f; }
+
+  void operator()(std::ptrdiff_t first, std::ptrdiff_t last) const {
+    ptrdiff_t len = last - first;
+    T* output_ptr = this->output + first; 
+        ConstEigenVectorArrayMap<T> xm(this->input + first, len);
+    EigenVectorArrayMap<T> ym(output_ptr, len);
+    ym = xm.log();
+  }
 };
 
 template <typename T>
-class Sub final : public OpKernel {
- public:
-  Sub(const OpKernelInfo& info) : OpKernel(info) {
+struct Abs : public ElementWiseRangedTransform<T> {
```

Comment:
Not sure the compiler will do much with it currently, but is there a reason to omit "final" compared with the prior version?

Reply:
ElementWiseRangedTransform is not derived class from OpKernel, it doesn't have a *prior* version. Anyway, i added final to it.